### PR TITLE
Implement 'Nil User' pattern to deal with deleted users

### PR DESCRIPTION
**Pull request by radar**
Potential fix for #427.

I think this is a good way to deal with deleted users. @knewter, @parndt, OK?

### CODE REVIEW

**Comment by parndt** [comment]
In general seems good.

**Comment by radar** [comment]
I've snuck in another change here which was required to fix some of the tests for the nil user.

It updates Forem to use Workflow 1.0.0 and switches some update_attribute calls to update_column calls so that the callbacks aren't triggered _all the damn time_. Workflow also provides the ability to define event methods which allows us to avoid using typical ActiveRecord callbacks for our state machine events.

File: spec/support/factories/topics.rb [hunk 1]
```diff
@@ -5,10 +5,10 @@
     t.user {|u| u.association(:user) }
     t.posts_attributes { [:text => "This is a brand new post"] }
 
-    trait :approved do
-      state 'approved'
+    factory :approved_topic do
+      after_create do |t|
+        t.approve!
```
Comment (radar) [inline spec/support/factories/topics.rb:10]: This is necessary to trigger the `approve` method callbacks.

File: app/views/forem/posts/_post.html.erb [hunk 2]
```diff
@@ -3,7 +3,11 @@
 
   <div class='user'>
     <div class='username'>
-      <%= link_to_if Forem.user_profile_links, post.user, [main_app, post.user] %>
+      <% if post.user.is_a?(Forem::NilUser) %>
+        <%= t(:deleted) %>
+      <% else %>
+        <%= link_to_if Forem.user_profile_links, post.user, [main_app, post.user] %>
+      <% end %>
```
Comment (parndt) [inline app/views/forem/posts/_post.html.erb:10]: Normally a NilUser would make the view less complicated ;-)

**Comment by avitus** [comment]
Would be great if this could be merged.

**Comment by parndt** [comment]
My only gripe was the view complication.

**Comment by radar** [comment]
Added to master and rails4. If anyone has any ideas on how to make the view less complicated, patches are welcome :)
